### PR TITLE
Fix #11833 and simplify injectdestructors

### LINKED ISSUE #11833: Newruntime: Double destruction of seq/string
The following destroys the inner seq twice with `--newruntime`, causing unexpected behavior. To have it immediately crash consistently, use `-d:useMalloc`.
(The `Foo` is not important. This happens with any non-constant element)

```nim
type Foo = object

proc test() =
  var x = @[@[Foo()]]

test()
```
From the generated code:
```cpp
T2_.len = 1; T2_.p = (tySequence_YBwSFr9cHnwXey9c6gHCsmuA_Content*) newSeqPayload(1, sizeof(tySequence_9aBObKOIQyNbQpSLHF0GtHw));
T3_.len = 1; T3_.p = (tySequence_9aBObKOIQyNbQpSLHF0GtHw_Content*) newSeqPayload(1, sizeof(tyObject_Foo_pMBAR5nJd0PD9cv5J1Hnd9bA));
...
eqsink__vltVh8v0Jk9bJ9bP86AQ9bCzA(colontmpD_, T3_);
T2_.p->data[0] = colontmpD_;
eqsink__6II9babRqgYpp6uFdPL4n3w(x, T2_);
...
eqdestroy__qK268sotx3d7bmLMEDvQtQ(colontmpD_);
eqdestroy__EKI3QOkat5QRImLwai9aGKw(x);
```
The inner `seq` is first deallocated through the temporary, then again as part of the outer `seq`.

Related (but not fixing the issue): The generated destructors don't nil the payload/length of the sequences.

### PR DESCRIPTION
**Pull request by Clyybber**

### CODE REVIEW

File: compiler/injectdestructors.nim [hunk 1]
```diff
@@ -553,8 +510,6 @@ proc pArg(arg: PNode; c: var Con; isSink: bool): PNode =
           branch = copyNode(arg[i])
           branch.add pArgIfTyped(arg[i][0])
         result.add branch
-    elif isAnalysableFieldAccess(arg, c.owner) and isLastRead(arg, c):
```
Comment (Araq) [inline compiler/injectdestructors.nim]: Huh? Why remove this?

Comment (Clyybber) [reply compiler/injectdestructors.nim]: Its a duplicate. See some lines above

File: compiler/injectdestructors.nim [hunk 2]
```diff
@@ -390,20 +347,21 @@ proc getTemp(c: var Con; typ: PType; info: TLineInfo): PNode =
...
+template isExpression(n: PNode): bool =
+  (not isEmptyType(n.typ)) or (n.kind in nkLiterals + {nkNilLit})
 
-proc genMagicCall(n: PNode; c: var Con; magicname: string; m: TMagic): PNode =
-  result = newNodeI(nkCall, n.info)
-  result.add(newSymNode(createMagic(c.graph, magicname, m)))
-  result.add n
+template recurse(n, dest) =
+  for son in n:
+    dest.add if isExpression(son): pExpr(son, c)
```
Comment (Araq) [inline compiler/injectdestructors.nim]: This looks sloppy to me. You should know from the context whether to call ``pExpr`` or ``pStmt``. There should be 2 different ``recurse`` templates for this then.

Comment (Clyybber) [reply compiler/injectdestructors.nim]: Yeah, I agree.

**Comment by Clyybber** [comment]
Not sure where the test failures on 32-bit come from...

**Comment by Araq** [comment]
It fails with 

Error: unhandled exception: injectdestructors.nim(474, 11) `false` nkChckRange64 [AssertionError]

Which are 32bit specific node kinds.

**Comment by Araq** [comment]
But to be honest, I don't see anymore how this rewrite is clearer than the current version. Sorry, but I would have to reject this, even if the tests were green.

**Comment by Clyybber** [comment]
@Araq It is more explicit and includes the partially materialized arrays for proc arguments and discard. This explicitness also lead to me discovering a bug which is visible in the transformation of system.getEnv.
We also move string literals now.
This explicitness is an improvement over the current implementation.
The amount of code duplication is also reduced with the new recurse proc, while at the same time being more explicit/"safe".
Sorry for the formatting, I dont have a physical keyboard at hand.

**Comment by Araq** [comment]
Ok, in the meantime, the bugfix is also here: https://github.com/nim-lang/Nim/pull/12018

File: compiler/injectdestructors.nim [hunk 3]
```diff
@@ -507,203 +568,218 @@ proc pArg(arg: PNode; c: var Con; isSink: bool): PNode =
...
   # we handle in the 'case of' statement below:
   const movableNodeKinds = (nkCallKinds + {nkSym, nkTupleConstr, nkObjConstr,
                                            nkBracket, nkBracketExpr, nkNilLit})
 
-  template moveOrCopyIfTyped(riPart: PNode): PNode =
-    # typ is nil if we are in if/case expr branch with noreturn
-    if riPart.typ == nil: p(riPart, c)
-    else: moveOrCopy(dest, riPart, c)
-
+  #XXX: All these nkStmtList results will cause problems in recursive moveOrCopy calls
```
Comment (Araq) [inline compiler/injectdestructors.nim]: Is that a new problem?

File: tests/destructor/t11833.nim [hunk 4]
```diff
@@ -0,0 +1,12 @@
+discard """
```
Comment (Araq) [inline tests/destructor/t11833.nim]: I added this test to some existing test, so please remove it.

File: compiler/injectdestructors.nim (hunk 3 above)
Comment (Clyybber) [reply compiler/injectdestructors.nim]: Nope, just thought I'd put a XXX here, for future reference.

File: tests/destructor/t11833.nim (hunk 4 above)
Comment (Clyybber) [reply tests/destructor/t11833.nim]: Ok

**Comment by Clyybber** [comment]
Unrelated test failures..

**Comment by Clyybber** [comment]
@Araq ping

**Comment by Araq** [comment]
Since this causes regressions, I'm reverting it. The statement/expression split is outdated, I don't use it anymore for transformations, see https://github.com/nim-lang/Nim/pull/12168 for how I would do it. It currently fails every test though...

**Pull request merged**